### ISSUE TITLE
Stop recommending PhantomJS

### ISSUE THREAD

**Issue by japgolly** [scala-js/scala-js#1881]
Scala.JS recommends Node.JS for DOM-less testing, and PhantomJS for testing with DOM.

Over the past year, PhantomJS has been nothing but a source of trouble for me. There are 1600+ open issues on its project page, many are serious and have been open without activity for years. In fact it doesn't get much activity at all. There are a few open issues within Scala.JS too (#1555, #1457, #1702, more?).

We shouldn't be directing Scala.JS users to use PhantomJS. Some other solution should be discovered and recommended.

I'm guessing node/io.js + jsdom (_@sjrd | @gzm0, if you've tried this I'd love to hear your experiences_), or maybe something using `xvfb` as suggested [here](https://groups.google.com/forum/#!folder/Scala/scala-js/d7S56w2VUpw). No answer yet, but I think it should be recognised that PhantomJS doesn't really work, and we're going to be setting the community up for eventual and significant pain, if we find a better solution here.

**Comment by mdedetrich** [comment]
tbh, I don't think any solution would be better than `PhantomJS`. Emulating the DOM **correctly** is hard, every browser does it differently, and with stuff continuously moving in HTML5 land, its hard to be ontop of especially with a spec that isn't really set in stone.

I don't have an issue in recommending something that is better, I am just not sure if a better exists

**Comment by japgolly** [comment]
It's not just niche/advanced/edge features that PhantomJS can't handle. It doesn't even support focusing (`.focus()`, `.blur()`, `document.activeElement`). :disappointed:

**Comment by mdedetrich** [comment]
I don't doubt that at all, my main point is that I am not aware of a better `Phantom.js` existing. Doing this stuff correctly is ridiculously hard, and you can use the old "standards" argument as to why thats the case.

Also issue count isn't an objective measure of how stable/correct implementation is. Once any non trivial project gets to a certain size/number of users, you get that many issues (a significant amount of those issues also turn out to be non issues)

**Comment by japgolly** [comment]
I don't doubt that it's hard. I also don't care, nor do I think any of us should. Scala.JS users' goals aren't to fix PhantomJS or build an alternative, our goals are to write Scala.JS code and be able to test it. PhantomJS is very limited in that regard and by observation of its continual lack of activity, isn't going to get much better. We should search for alternatives. A suitable alternative *product* doesn't seem to exist AFAIK (i.e. a reliable, headless browser), but alternative *means* of achieving the same goal do exist, and its the exploration of those that I believe is in the best (long-term) interest of the community.

**Comment by japgolly** [comment]
Oh although node|io.js + jsdom might be a good alternate solution in the same vein, but I'm not sure that anyone's got it working with Scala.JS. I'm hoping to try myself, maybe this weekend if I get time. That has potential I think.

**Comment by benjaminjackman** [comment]
This is hard, because it goes beyond just libraries that are ostensibly dom-centric (e.g. d3). some libraries that are mostly not dom related still end up requiring it. I had to use phantom because moment.js (a date library) was requiring the dom for some reason and without it was blowing up on unit tests. This was a while ago, so maybe it's not an issue anymore with this specific. The problem is that a lot of random js code that can end up in a project might have a few lines that refer some dom api without which they just blow up.

I guess overall the best thing is to, in addition to what is suggested here, recommend a differentation at the project level between portions of ScalaJS the user writes that interacts with the DOM and pieces that don't.

This means there is almost kind of a hierarchy:

```
1> Platform-Agostic Scala (The portions of scala ecosystem / the jvm APIs which work in ScalaJS)|
|2> JVM-Only Scala (uses scala/java libraries / apis not ported / supported to ScalaJS)
|3> JS-Only Scala (uses things that are ScalaJS specific but which should run on any reasonable javascript runtime, e.g. js.Dynamic.literal(), maybe some work should be done here to port this code over to work on the JVM at some point, i dunno kind of bends my mind just thinking about that)
_|4>  Node JS Scala (uses node apis) runs on server (and emulates dom with Phantom or something better)
_|5>  DOM JS Scala (uses dom apis) runs in browser (and emulates node with browserify)
```
So from my example, I expected moment.js to be a `3` type of library whereas really it turned out to be a `5` type.

I think the testing story for the `5` type libraries / code is always going to be tougher than for 1-4. So for things like moment which are `5`'s but which should really be `3` something like phantom (or equivalent might be ok if you can get it to run on your platform and it still will crash the tests for whatever reason once in a while #1457 et al.), 

For `5`s that are a bunch of real DOM-modifying UI browsery goodness, I think you have to go to selenium / firefox aka actual in browser testing if you want anything reasonable from what I hear (but I have not done this myself yet). 

So maybe we should suggest, users try to write type `1`(write once, run everywhere) code if at all possible, then type `3` (no dom/node agnostic scalajs) code. If they are doing actual DOM / UI testing phantom is buggy, and they should use selenium, and then add some SBT plugins (or patch the SBT ScalaJS plugin) that would get firefox / selenium testing working?

**Comment by easel** [comment]
A couple thoughts:

1. I'm not sure headless testing is really the goal, so much as the ability to reliably unit test code that will run in a browser environment. Ideally, this can be done in a browser agnostic way. If that problem is solved, making it truly headless can also be solved by introducing xvfb or falling back to phantomjs.

2. While node+jsdom will inevitably mature and improve, it's never going to be a real browser. Given the jvm, most scala.js code will be deployed to browsers for the foreseeable future. For sufficiently limited code, node and node+jsdom will provide a nice optimization, but I think it's important to target real browsers and then optimize down from there rather than the other way.

3. Selenium/webdriver isn't that relevant to unit testing. Web driver provides a json api for actually controlling the browser, so you can use it to click the back button, fill out forms, etc. As such, it requires direct support from the browser manufacturer or a plugin, and provides a bunch of functionality that isn't really relevant to scala.js so much as for writing functional tests. It *may* be the most appropriate way to start the browser and feed unit test results back, but it's not the only way and it's a finicky one so I don't think designing a unit test solution around it exclusively is a good idea.

That said, the javascript community has more or less solved this problem. They push an html page with javascript in it to the browser, render it, and display the results in the page. Tools like karma capture these results, allowing for a command line interface, CI, etc. This is simple, browser agnostic, and works remotely .

I believe scala.js already has some pieces of this in place. There are jasmine tests for the core and the phantomjs test runner seems to generate some html code for the test runner. There's also a jetty process wired into sbt to support phantomjs, so it feels to me like these things could be reorganized in such a way that you could use the browser of your choice instead of just phantomjs.

**Comment by gzm0** [comment]
I think we can officially consider this fixed with #2583.
